src/clp_ffi_js/ir/StructuredIrUnitHandler.cpp

```diff
@@ -0,0 +1,157 @@
+namespace {
+/**
+ * Parses a string to determine the corresponding `LogLevel` enum value.
+ * @param str
+ * @return `LogLevel` enum corresponding to `str` if `str` matches a string in `cLogLevelNames`.
+ * @return `LogLevel::NONE` otherwise.
+ */
+auto parse_log_level(std::string_view str) -> LogLevel;
+
+auto parse_log_level(std::string_view str) -> LogLevel {
+    // Convert the string to uppercase.
+    std::string log_level_name_upper_case{str};
+    std::ranges::transform(
+            log_level_name_upper_case.begin(),
+            log_level_name_upper_case.end(),
+            log_level_name_upper_case.begin(),
+            [](unsigned char c) { return std::toupper(c); }
+    );
+
+    auto const* it = std::ranges::find(
+            cLogLevelNames.begin() + clp::enum_to_underlying_type(cValidLogLevelsBeginIdx),
+            cLogLevelNames.end(),
+            log_level_name_upper_case
+    );
+    if (it == cLogLevelNames.end()) {
+        return LogLevel::NONE;
+    }
+
+    return static_cast<LogLevel>(std::distance(cLogLevelNames.begin(), it));
+}
+}  // namespace
+
+auto StructuredIrUnitHandler::handle_log_event(StructuredLogEvent&& log_event
+) -> clp::ffi::ir_stream::IRErrorCode {
+    auto const& id_value_pairs{log_event.get_node_id_value_pairs()};
+    auto const timestamp = get_timestamp(id_value_pairs);
+    auto const log_level = get_log_level(id_value_pairs);
+
+    m_deserialized_log_events->emplace_back(std::move(log_event), log_level, timestamp);
+
+    return clp::ffi::ir_stream::IRErrorCode::IRErrorCode_Success;
+}
+
+auto StructuredIrUnitHandler::handle_utc_offset_change(
+        [[maybe_unused]] clp::UtcOffset utc_offset_old,
+        [[maybe_unused]] clp::UtcOffset utc_offset_new
+) -> clp::ffi::ir_stream::IRErrorCode {
+    SPDLOG_WARN("UTC offset change packets aren't handled currently.");
+    return clp::ffi::ir_stream::IRErrorCode::IRErrorCode_Success;
+}
+
+auto StructuredIrUnitHandler::handle_schema_tree_node_insertion(
+        clp::ffi::SchemaTree::NodeLocator schema_tree_node_locator
+) -> clp::ffi::ir_stream::IRErrorCode {
+    ++m_current_node_id;
+
+    auto const& key_name{schema_tree_node_locator.get_key_name()};
+    if (key_name == m_log_level_key) {
+        m_log_level_node_id.emplace(m_current_node_id);
+    } else if (key_name == m_timestamp_key) {
+        m_timestamp_node_id.emplace(m_current_node_id);
+    }
+
+    return clp::ffi::ir_stream::IRErrorCode::IRErrorCode_Success;
+}
+
+auto StructuredIrUnitHandler::handle_end_of_stream() -> clp::ffi::ir_stream::IRErrorCode {
+    return clp::ffi::ir_stream::IRErrorCode::IRErrorCode_Success;
+}
+
+auto StructuredIrUnitHandler::get_log_level(
+        StructuredLogEvent::NodeIdValuePairs const& id_value_pairs
+) const -> LogLevel {
+    LogLevel log_level{LogLevel::NONE};
+
+    if (false == m_log_level_node_id.has_value()) {
+        return log_level;
+    }
+    auto const& log_level_pair{id_value_pairs.at(m_log_level_node_id.value())};
```

Comment:
I thought I posted this in the previous review but it's not there so I guess I didn't post it or GH lost it before posting. Anyway-

This isn't a pair, right? This is the value from the `id_value_pairs` entry. Since the value itself is an optional, how about `optional_log_level_value`? Then after checking that the optional isn't empty, we can add a variable `log_level_value` and use that in the rest of the method.

We can do the same thing in `get_timestamp`.